### ISSUE TITLE
CNN on Speech Commands Dataset

### ISSUE THREAD

**Issue by EdgarLemaire** [CEA-LIST/N2D2#72]
Hello, I believe that N2D2 should work with Speech Commands Dataset, as it is presented as a possible custom dataset in the N2D2 manual.
However, when trying with a CNN, I cannot perform a classification accuracy above 10%. Do you have a ini file example for Speech Commands ?
Thank you,

Edgar

**Comment by olivierbichler-cea** [comment]
Hi,

I don't have any example available, but you may post your INI file for checking!

Cheers,
Olivier

**Comment by EdgarLemaire** [comment]
Hello Olivier,
Thanks for the quick answer.


First of all, note that I have pre-processed (Mel Spectrogram) the data with pytorch. After pre-processing, the data is of shape (101,40,3). Instead of feeding it to the network (which acheives 93% on pytorch with the exact same topology), I save it as RGB PNG images. I have organized the dataset with one folder per class as required by N2D2.

Here is my INI file  :


--------------------------------------------------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------


DefaultModel=Frame_CUDA



; Database
[database]
Type=DIR_Database
DataPath=/home/edlemaire/N2D2_DATA/Speech_Commands
Learn=0.8
Validation=0.1 ; Use 20% of the dataset for validation

; Environment
[sp]
SizeX=40
SizeY=101
NbChannels=3


[conv1]
Input=sp
Type=Conv
KernelDims=7 7
NbOutputs=32
ActivationFunction=Rectifier
WeightsFiller=HeFiller
ConfigSection=common.config

[pool1]
Input=conv1
Type=Pool
PoolDims=2 2
NbOutputs=[conv1]NbOutputs
Stride=2
Pooling=Max
Mapping.Size=1

[conv2]
Input=conv1
Type=Conv
KernelDims=5 5
NbOutputs=32
ActivationFunction=Rectifier
WeightsFiller=HeFiller
ConfigSection=common.config

[pool2]
Input=conv2
Type=Pool
PoolDims=2 2
NbOutputs=[conv2]NbOutputs
Stride=2
Pooling=Max
Mapping.Size=1

[conv3]
Input=pool2
Type=Conv
KernelDims=5 5
NbOutputs=32
ActivationFunction=Rectifier
WeightsFiller=HeFiller
ConfigSection=common.config

[pool3]
Input=conv3
Type=Pool
PoolDims=2 2
NbOutputs=[conv3]NbOutputs
Stride=2
Pooling=Max
Mapping.Size=1


[fc1]
Input=pool3
Type=Fc
NbOutputs=12
ActivationFunction=Linear
ConfigSection=common.config

[softmax]
Input=fc1
Type=Softmax
NbOutputs=[fc1]NbOutputs
WithLoss=1

[softmax.Target]

[common.config]
NoBias=1
WeightsSolver.LearningRate=0.05
WeightsSolver.Momentum=0.9
WeightsSolver.Decay=0.0005
Solvers.LearningRatePolicy=StepDecay
Solvers.LearningRateStepSize=[sp]_EpochSize
Solvers.LearningRateDecay=0.8
Solvers.Clamping=-1.0:1.0

**Comment by EdgarLemaire** [comment]
I have also tried with the raw Speech Command Dataset (without Mel Spectrogram transformation), with the same topology, without more success...

**Comment by olivierbichler-cea** [comment]
Hello,

Shouldn't the input of conv2 be pool1 instead of conv1?

Olivier

**Comment by EdgarLemaire** [comment]
Hello Olivier,

Yes you are right. However the problem was not here, it suddenly acheived SOTA classification score without changing anything. I guess this has something to do with the weight initialization.

Regards,

Edgar

**Comment by olivierbichler-cea** [comment]
Could you be more specific? Do you mean the simulation in N2D2 now reaches the same accuracy than in PyTorch?
If it works as intended, I will close this issue.
Cheers,
Olivier

**Comment by EdgarLemaire** [comment]
Yes, simulation in N2D2 reaches the same accuracy as in PyTorch, but not on every run. In the majority of runs, the accuracy is far from reaching the PyTorch level. That is why I guess it has something to do with initialization.
However, I think you can close the issue :)

Have you seen my other issue concerning DeepSig by the way ? I have a similar problem and cannot resolve it this time...

**Comment by olivierbichler-cea** [comment]
Do you know how the weights are initialized in PyTorch? I am sure we can configure the INI to be equivalent!

**Comment by EdgarLemaire** [comment]
Yes, they are initialized following a normal distribution with mean 0 and stddev 0.15

**Comment by olivierbichler-cea** [comment]
Ok, so in N2D2 you just need to add the following in the layer sections:

```
WeightsFiller=NormalFiller
WeightsFiller.Mean=0.0
WeightsFiller.StdDev=0.15
```

**Comment by EdgarLemaire** [comment]
Hello Olivier,

Thank you for the parameters, I didn't know such options could be used in N2D2. It seems to work better like this :-)

You can close the issue now,


Edgar
